### ISSUE TITLE
Kokkos Kernels fails to build with LayoutLeft=OFF

### ISSUE THREAD

**Issue by lucbv** [kokkos/kokkos-kernels#903]
@ndellingwood @brian-kelley @vqd8a <--- not too sure who would know about this?
I get build failures in various locations when forcing `Kokkos::LayoutRight` only by enabling LAYOUTRIGHT and disabling LAYOUTLEFT, see CMake configuration below used on Kokkos-dev2 with Intel 19.0.5:
```
cmake -D CMAKE_CXX_COMPILER:STRING="icpc" \
      -D CMAKE_CXX_STANDARD:STRING="14" \
      -D CMAKE_BUILD_TYPE:STRING="RelWithDebInfo" \
      -D Kokkos_DIR:PATH="${KOKKOS_INSTALL}/lib64/cmake/Kokkos" \
      -D CMAKE_INSTALL_PREFIX:PATH="${KOKKOSKERNELS_INSTALL}" \
      -D KokkosKernels_ENABLE_TESTS:BOOL=ON \
      -D KokkosKernels_ENABLE_EXAMPLES:BOOL=ON \
      -D KokkosKernels_INST_LAYOUTLEFT:BOOL=OFF \
      -D KokkosKernels_INST_LAYOUTRIGHT:BOOL=ON \
      -D KokkosKernels_INST_DOUBLE:BOOL=ON \
      -D KokkosKernels_INST_COMPLEX_DOUBLE:BOOL=ON \
      -D KokkosKernels_INST_ORDINAL_INT:BOOL=ON \
      -D KokkosKernels_INST_OFFSET_INT:BOOL=ON \
      ${KOKKOSKERNELS_PATH}
```
The errors I observe are of the following nature:
```
CMakeFiles/KokkosKernels_blas_openmp.dir/openmp/Test_OpenMP_Blas1_abs.cpp.o: In function `void KokkosBlas::abs<Kokkos::View<double*, Kokkos::LayoutRight, Kokkos::OpenMP>, Kokkos::View<double const*, Kokkos::La\
youtRight, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0u> > >(Kokkos::View<double*, Kokkos::LayoutRight, Kokkos::OpenMP> const&, Kokkos::View<double const*, Kokkos::LayoutRight, Ko\
kkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<0u> > const&)':
/home/lberge/kk_lucbv/src/blas/KokkosBlas1_abs.hpp:112: undefined reference to `KokkosBlas::Impl::Abs<Kokkos::View<double*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::Memory\
Traits<1u> >, Kokkos::View<double const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> >, 1, false, false>::abs(Kokkos::View<double*, Kokkos::LayoutLeft, Kokk\
os::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> > const&, Kokkos::View<double const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::MemoryTraits<1u> > c\
onst&)'
```
It looks like something wrong is going on in terms of ETI or implicit instantiation... is the CMake configuration above invalid for some reason and should it fail at configure time or do we have something wrong in the code?

**Comment by e10harvey** [comment]
@lucbv Thanks for reporting this. I too have been seeing link errors when trying to build with only `LayoutRight`. Are `KOKKOSKERNELS_ETI_ONLY` and `KOKKOSKERNELS_IMPL_CHECK_ETI_CALLS` undefined? I see that this could cause LayoutLeft to be passed in via:
```C++
template<class ScalarA, class ScalarB, class Device>
int test_abs() {

#if defined(KOKKOSKERNELS_INST_LAYOUTLEFT) || (!defined(KOKKOSKERNELS_ETI_ONLY) && !defined(KOKKOSKERNELS_IMPL_CHECK_ETI_CALLS))
  typedef Kokkos::View<ScalarA*, Kokkos::LayoutLeft, Device> view_type_a_ll;
```

**Comment by vqd8a** [comment]
@lucbv I saw these link errors last week, but not sure what caused them. 
@e10harvey, I also saw that these errors related to half precision. Could you please take a look? Thanks.

**Comment by e10harvey** [comment]
> @e10harvey, I also saw that these errors related to half precision. Could you please take a look? Thanks.

@vqd8a Please open an issue with a reproducer for this and assign it to me.

**Comment by vqd8a** [comment]
@e10harvey I have tried building Kokkos Kernels again and do see the same link errors with ```Kokkos::LayoutRight```. I remembered wrong about the errors with half precision. I got the errors with half precision not with your branch ```perf_test_updates```, not ```develop``` branch. Please ignore my previous comment. I can talk with you about the errors with half precision later.

**Comment by brian-kelley** [comment]
@lucbv This one isn't caused by an ETI bug, it's just that the KokkosKernels tests don't allow instantiating for non-ETI types. All the blas functions convert their rank-1 inputs to unmanaged LayoutLeft (via GetUnifiedLayout) and so the blas impls are all getting LayoutLeft. I think we could safely change GetUnifiedLayout to produce rank-0/1 LayoutRight views if that's the only one that's enabled in ETI.

**Comment by e10harvey** [comment]
Addressed via #965.
